test/integration/segmentation_camera_plugin.cc

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <gtest/gtest.h>
+
+#include <ignition/common/Filesystem.hh>
+#include <ignition/sensors/Manager.hh>
+#include <ignition/sensors/SegmentationCameraSensor.hh>
+
+#ifdef _WIN32
+#pragma warning(push)
+#pragma warning(disable: 4251)
+#endif
+#include <ignition/rendering/RenderEngine.hh>
+#include <ignition/rendering/RenderingIface.hh>
+#include <ignition/rendering/Scene.hh>
+#include <ignition/rendering/SegmentationCamera.hh>
+#ifdef _WIN32
+#pragma warning(pop)
+#endif
+
+#ifdef _WIN32
+#pragma warning(push)
+#pragma warning(disable: 4005)
+#pragma warning(disable: 4251)
+#endif
+#include <ignition/msgs.hh>
+#ifdef _WIN32
+#pragma warning(pop)
+#endif
+
+#include "test_config.h"  // NOLINT(build/include)
+#include "TransportTestTools.hh"
+
+using namespace ignition;
+using namespace rendering;
+
+class SegmentationCameraSensorTest: public testing::Test,
+  public testing::WithParamInterface<const char *>
+{
+  // Create a Segmentation Camera sensor from a SDF and gets a image message
+  public: void ImagesWithBuiltinSDF(const std::string &_renderEngine);
+};
+
+/// \brief mutex for thread safety
+std::mutex g_mutex;
+
+/// \brief Segmentation buffer
+unsigned char *g_buffer = nullptr;
+
+/// \brief counter of received segmentation msgs
+int g_counter = 0;
+
+/// \brief callback to get the segmentation buffer
+void OnNewSegmentationFrame(const msgs::Image &_msg)
+{
+  g_mutex.lock();
+  unsigned int size = _msg.width() * _msg.height() * 3;
+
+  if (!g_buffer)
+    g_buffer = new unsigned char[size];
+
+  memcpy(g_buffer, _msg.data().c_str(), size);
+  g_counter++;
+  g_mutex.unlock();
+}
+
+/// \brief wait till you read the published frame
+void WaitForNewFrame()
+{
+  auto waitTime = std::chrono::duration_cast< std::chrono::milliseconds >(
+  std::chrono::duration< double >(0.001));
+  int counter = 0;
+  // wait for the counter to increase
+  for (int sleep = 0; sleep < 300 && counter == 0; ++sleep)
+  {
+    g_mutex.lock();
+    counter = g_counter;
+    g_mutex.unlock();
+    std::this_thread::sleep_for(waitTime);
+  }
+  EXPECT_GT(counter, 0);
+}
```

Comment:
My concern with this method is that it doesn't actually tell us if we received an image or not - it will tell us if `g_counter` isn't zero, but it won't tell us if `g_counter` has changed (if `g_counter` is already > 0 when this method is called, then this method doesn't do us any good). I recommend making the following changes: pass in the `ignition::sensors::Manager` as an argument, set `g_counter` to 0, and THEN call `mgr.RunOnce` in the `WaitForNewFrame` method. That way, when the `EXPECT_GT(counter, 0)` call is reached at the end of the method, that lets us know if an image was actually received or not. Once you make these changes, you can modify the actual test method to call just `WaitForNewFrame(mgr)` instead of calling `mgr.RunOnce` followed by `WaitForNewFrame()`.